### PR TITLE
Add traversal driver.

### PR DESCRIPTION
**Pull request by jone**
The traversal driver executes the requests by traversing internally in the same transaction and with the same connection as the test.

The advantage of the traversal driver is that it does not need uncommitted state to be committed, nor does it commit modifications happening in the request.

// cc @4teamwork/gever @buchi @maethu

### CODE REVIEW

File: ftw/testbrowser/drivers/traversaldriver.py [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+    """Wraps a Zope2.App.startup.TransactionsManager and patches
+    Zope2.zpublisher_transactions_manager with a singleton instance.
+
+    The wrapper lives for the rest of the process duration since it
+    is cached in the function argument default dict ``modules`` of the
+    function ``ZPublisher.Publish.get_module_info``.
+
+    When the wrapper is activated with ``activate_wrapper`` it silently
+    aborts all commit intents.
+    When then transaction is beginning, we replace the ZApplicationWrapper
```
Comment (deiferni) [inline ftw/testbrowser/drivers/traversaldriver.py]: maybe rephrase this a bit, and add more background info?

File: ftw/testbrowser/drivers/traversaldriver.py [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+
+    def abort(self):
+        if not self.active:
+            return self.wrapped.abort()
+
+    def recordMetaData(self, object, request):
+        if not self.active:
+            return self.wrapped.recordMetaData(object, request)
+
+    def set_app_in_parents(self):
```
Comment (deiferni) [inline ftw/testbrowser/drivers/traversaldriver.py]: a bit unhappy about the name, maybe `inject_into_parents` or similar would be more clear?

File: ftw/testbrowser/drivers/traversaldriver.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+        """When the ``ZApplicationWrapper`` is in the ``PARENTS`` list,
+        it will make a new DB connection when bobo-traversed, which will
+        not include the uncomitted state of the current transaction.
```
Comment (deiferni) [inline ftw/testbrowser/drivers/traversaldriver.py]: unco**m**mitted

File: ftw/testbrowser/drivers/traversaldriver.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+
+    def __init__(self, wrapped):
+        self.wrapped = wrapped
+        self.active = False
+        self.next_request = None
+        self.current_app = None
+
+    def activate_wrapper(self, next_request, current_app):
+        self.active = True
+        self.next_request = next_request
```
Comment (deiferni) [inline ftw/testbrowser/drivers/traversaldriver.py]: i don't really get what `next_request` and `current_app` are here for 😢 , could you add a docstring?

File: ftw/testbrowser/drivers/traversaldriver.py (hunk 1 above)
Comment (jone) [inline ftw/testbrowser/drivers/traversaldriver.py]: I have rewritten this docstring.

File: ftw/testbrowser/drivers/traversaldriver.py (hunk 4 above)
Comment (jone) [inline ftw/testbrowser/drivers/traversaldriver.py]: I have realized that this is actually a context manager and changed it accordingly.
This is now done in ``__call__``, I have added a docstring.

File: ftw/testbrowser/drivers/traversaldriver.py (hunk 2 above)
Comment (jone) [inline ftw/testbrowser/drivers/traversaldriver.py]: changed to `replace_app_in_parents` and simplified (removed the other cases which actually do not occur)

**Comment by jone** [comment]
@deiferni I've updated the docstrings and comments so that it should be clearer what we are doing here. I also have refactored a bit (use context manager, rename things, remove dead code).
I committed separately but will squash before merge.
